RecoTauTag/RecoTau/src/RecoTauDiscriminantFunctions.cc

```diff
@@ -20,16 +21,15 @@ const RecoTauPiZero& removeRef(const RecoTauPiZero& piZero) {
   return piZero;
 }
 
-// Function = A PFTau member function
-template<typename Collection, typename Function>
-VDouble extract(const Collection& cands, Function func) {
-  // #define CALL_MEMBER_FN(object,ptrToMember)  ((object).*(ptrToMember))
-  VDouble output;
-  output.reserve(cands.size());
-  for(typename Collection::const_iterator cand = cands.begin();
-      cand != cands.end(); ++cand) {
-    output.push_back(func(removeRef(*cand)));
-  }
+// Function = A CandidateType member function or any function taking a CandidateType
+template<typename T, typename Function>
+VDouble extract(const std::vector<T>& cands, Function function) {
+  VDouble output( cands.size() );
+  using CandidateType = decltype(removeRef(std::declval<T>()));
+  // Wrap function in std::funtion such that member function pointers and other
+  // functions (lambdas, normal functions, etc.) can be used in the same way
+  std::function<double(CandidateType const&)> stdFunction{ function };
+  for(auto const& x : cands) output.push_back(stdFunction(removeRef(x)));
```

Comment:
does ```std::decay``` simplify this at all?

Reply:
I'm not familiar with that one and didn't understand immediately how it would help here even after googling. You think it could replace this `removeRef` helper, or are you talking about generalizing for member functions and free functions?

Reply:
Anyway I'll experiment a bit

Reply:
I didn't have an explicit solution in mind.
Given most recent updates, perhaps https://en.cppreference.com/w/cpp/types/remove_pointer is more inspirational

Reply:
Ah yes, this also exists! But it does not work with edm::Ptr I imagine.

Reply:
ok, admittedly, my last comment is going a bit in the backward direction from what is in the PR now and implicitly suggests adding back an equivalent of ```removeRef(x)``` which returns the same for bare ```T``` and ```Ptr<T>```.
It's more practical to add this adaptor than to reimplement the body of the loop